File: libs/message_pool.js
class RingBuffer {
  constructor ( size ) {
    this._length = 0;
    this._start = 0;
    this._size = size
    this._data = new Array(size)
    this._end  = size -1
    this._overflow_cb = null; // callback function, called when overflow is occurred

    this.pop     = this.get_back
    this.shift   = this.get_front
    this.push    = this.add_back
    //this.unshift = this.add_front
  }
  get size() {
    return this._size;
  }
  get length() {
    return this._length;
  }
  index(idx) {
    if (idx >= this._length) {
      return undefined;
    }
    if (idx < 0) {
      if( this._length + idx < 0) {
        return undefined;
      }
      idx += this._length;
    }
    const nidx = (this._start + idx) % this._size;
    return this._data[nidx];
  }
  get_back () {
    //
    if (this._length === 0) return;

    //
    var elm = this._data[this._end];

    //
    delete this._data[this._end];

    this._end = (this._end - 1 + this._size) % this._size;
    this._length--;

    return elm;
  }
  get_front() {
    //
    if (this._length === 0) return;

    //
    var elm = this._data[this._start];

    //
    delete this._data[this._start];

    this._start = (this._start + 1) % this._size;
    this._length--;

    return elm;
  }
  _get_n(n, func) {
    if (n <= 0) {
      return;
    }
    var res = [];
    for(var i = 0; i < n; i++) {
      res.push( func.apply() )
    }
    return res;
  }
  get_backn (n) {
    return this._get_n(n, this.get_back.bind(this))// callback as method
  }
  get_frontn (n) {
    return this._get_n(n, this.get_front.bind(this))// callback as method
  }
  _add_back (elm) { // add_back for single item
    var new_end = this._end = (this._end + 1) % this._size;

    // if overflow_cb is set, and if data is about to be overwritten
    if (this._overflow_cb && this._length + 1 > this._size) {
      this._overflow_cb( this._data[new_end], this );
    }
    this._data[ new_end ] = elm;

    // recalculate length
    if (this._length < this._size) {
      if (this._length + 1 > this._size) {
        this._length = this._size;
      } else {
        this._length += 1;
      }
    }
    this._end = new_end;
    this._start = (this._size + this._end - this._length + 1) % this._size;
  }
  add_back() {
    if (arguments.length < 0) {
      return this.length;
    }
    for (var i = 0; i < arguments.length; i++) {
      this._add_back(arguments[i])
    }
    return this._length;
  }
  _add_front(elm) { // add_front for single item
    // if overflow_cb is set, and if data is about to be overwritten
    if (this._overflow_cb && this._length + 1 > this._size) {
      this._overflow_cb( this._data[ this._end ], this);
    }
    this._data[ (this._size + this._start - 1) % this._size ] = elm;

    // recalculate size
    if (this._size - this._length - 1 < 0) {
      this._end += this._size - this._length - 1;
      if (this._end < 0) {
        this._end = this._size + (this._end % this._size);
      }
    }
    // recalculate length
    if (this._length < this._size) {
      if (this._length + 1 > this._size) {
        this._length = this._size;
      } else {
        this._length += 1;
      }
    }
    this._start -= 1;
    if (this._start < 0) {
      this._start = this._size + (this._start % this._size);
    }
  }
  add_front() {
    if (arguments.length < 0) {
      return this.length;
    }
    for (var i = 0; i < arguments.length; i++) {
      this._add_front(argumehts[i])
    }
    return this._length;
  }
  unshift() {
    if (arguments.length < 0) {
      return this.length;
    }
    // for (var i = 0; i < arguments.length; i++) {
    for (var i = arguments.length - 1; i >= 0; i--) { // reverse order | same as javascript array
      this._add_front(argumehts[i])
    }
    return this._length;
  }
  get_all() {
    // not implemented
  }
  clear() {
    // this._size = size // no resize
    this._length = 0;
    this._start  = 0;
    this._data = new Array(this._size)
    this._end  = this._size -1
  }
};

class TopicBuffer extends RingBuffer
{
  constructor (size, callback = null )
  {
    super(size);
    this.callback = callback; // callback := function (msg) ...
    // if a callback function returns 'true', a message is not stored after callback called
  }
};

class MessagePool {
//exports.MessagePool = class {
  constructor (_size = 100)
  {
    this.ws_client = null;
    this.topic_map = new Map();
    this.buffer_size = _size;
    this.publish_method = false;
    this.debug = false;

    this.auto_subscribe = false;
    this.default_callback = null;
  }
  setPublishMethod(func)
  {
    this.publish_method = func;
  }
  static makeMessageStr(topic_name, data)
  {
    const msg = {
      hash: "TUTCPS",
      topic: topic_name,
      data:  data
    };
    const msg_str = JSON.stringify(msg);
    return msg_str;
  }
  publish(topic_name, data)//"topic_name", data)
  {
    if (this.publish_method) {
      const msg_str = MessagePool.makeMessageStr(topic_name, data);
      if (this.debug) {
        console.log('publish: ' + msg_str);
      }
      this.publish_method(msg_str);
    }
  }
  getMessage(topic_name, timeout = -1)// ("topic_name", timeout) => time, data
  {
    if ( this.topic_map.has(topic_name) ) {
      var buf = this.topic_map.get(topic_name);
      return buf.get_front()
    }
  }
  readLastMessage(topic_name, clear = false)// read last message
  {
    if ( this.topic_map.has(topic_name) ) {
      var buf = this.topic_map.get(topic_name);
      var ret = buf.index(-1);
      if (clear) {
        buf.clear();
      }
      return ret;
    }
  }
  subscribe(topic_name, callback = null)//("topic_name", function=None)
  {
    if ( this.topic_map.has(topic_name) ) {
      // already subscribed
      var buf = this.topic_map.get(topic_name)
      if (callback) {
        if (buf.callback) {
          // overwrite callback
          buf.callback = callback
        } else {
          // add callback
          buf.callback = callback
        }
      }
    } else {
      this.topic_map.set(topic_name, new TopicBuffer(this.buffer_size, callback))
    }
  }
  unsubscribe(topic_name)
  {
    if ( this.topic_map.has(topic_name) ) {
      this.topic_map.delete(topic_name);
    }
  }
  // attribute
  get subscribingTopicList()
  {
    return Array.from(this.topic_map.keys())
  }
  get subscribingTopicIter()
  {
    return this.topic_map.keys();
  }
  //// private method
  _receiveData(instr) // when messageRecv
  {
    //parsing -> run callback -> store message
    var msg = JSON.parse(instr);
    if (this.debug) {
      console.log("revc: " + String(msg));
    }
    if (msg.hash != undefined && msg.hash == "TUTCPS") {
      if (msg.topic != undefined && msg.data != undefined) {
        if ( this.topic_map.has(msg.topic) ) {
          var buf = this.topic_map.get(msg.topic);
          if (this.debug) {
            console.log("cb_func: " + buf.callback);
          }
          //if (buf.callback && buf.callback.apply(buf, [msg.data])) {
          if (buf.callback && buf.callback(msg.data)) {
            if (this.debug) {
              console.log('do not stack message to buffer');
            }
            return;
          }
          if (this.debug) {
            console.log('stack message to buffer');
          }
          buf.push([new Date(), msg.data]);
        } else {
          if (this.debug) {
            console.log('receive not subscribing topic : ' + msg.topic);
          }
          if (this.auto_subscribe) {
            this.subscribe(msg.topic, this.default_callback);
            this._receiveData(instr);
          }
          // receive not subscribing data
          //buf = new RingBuffer(this.buffer_size);
          //this.topic_map.set(msg.topic,  buf);
        }
      }
    }
  }
};

class WebMessage {
  constructor(addr, _size = 100, _auth = false)
  {
    this.addr = addr;
    this.size = _size;
    this.authorized = _auth;
    this.connection = false;
    this.connect();
  }
  connect()
  {
    this.ws_client = new WebSocket(this.addr, null, null, null, {rejectUnauthorized: this.authorized});
    this.msg_pool  = new MessagePool(this.size);
    const _wsc = this.ws_client;
    this.msg_pool.setPublishMethod( (instr) => {
      _wsc.send(instr);
    } );

    const _pool = this.msg_pool;
    this.ws_client.onmessage = (e) => {
      // DEBUG
      //console.log('recv message');//debug
      //console.log(e.data);//debug
      _pool._receiveData(e.data);
    };
    this.ws_client.onopen = (e) => {
      console.log('ws open');
      // write header
    };
    this.ws_client.onerror = (e) => {
      console.log('ws error');
      // write header
    };
    this.ws_client.onclose = (e) => {
      console.log('ws closed');
      // write header
    };
  }
};

class BrowserMessage {
  constructor(hostwindow, _size = 100)
  {
    this.size = _size;
    this.hostwindow = hostwindow;
    this.connection = false;
    this.connect();
  }
  connect ()
  {
    this.msg_pool = new MessagePool(this.size);
    var pool = this.msg_pool;
    var host = this.hostwindow;
    this.msg_pool.setPublishMethod((instr) => {
      host.postMessage(instr, '*');
    });
    window.addEventListener("message", //// message from host
                            (ev) => {
                              //console.log('msg:' + ev.data);
                              pool._receiveData(ev.data);
                            }, false);
    this.connection = true;
  }
};

File: python/message_pool.py
#!/usr/bin/env python3
"""Message pool utilities translated from message_pool.js."""

from __future__ import annotations

import json
from datetime import datetime
from typing import Any, Callable, Dict, Iterable, Iterator, List, Optional, Set

#
#
#
class RingBuffer:
    def __init__(self, size: int) -> None:
        self._length = 0
        self._start = 0
        self._size = size
        self._data: List[Any] = [None] * size
        self._end = size - 1
        self._overflow_cb: Optional[Callable[[Any, "RingBuffer"], None]] = None

    @property
    def size(self) -> int:
        return self._size

    @property
    def length(self) -> int:
        return self._length

    def index(self, idx: int) -> Any:
        if idx >= self._length:
            return None
        if idx < 0:
            if self._length + idx < 0:
                return None
            idx += self._length
        nidx = (self._start + idx) % self._size
        return self._data[nidx]

    def pop(self) -> Any:
        return self.get_back()

    def shift(self) -> Any:
        return self.get_front()

    def push(self, *items: Any) -> int:
        return self.add_back(*items)

    def get_back(self) -> Any:
        if self._length == 0:
            return None

        elm = self._data[self._end]
        self._data[self._end] = None
        self._end = (self._end - 1 + self._size) % self._size
        self._length -= 1
        return elm

    def get_front(self) -> Any:
        if self._length == 0:
            return None

        elm = self._data[self._start]
        self._data[self._start] = None
        self._start = (self._start + 1) % self._size
        self._length -= 1
        return elm

    def _get_n(self, n: int, func: Callable[[], Any]) -> Optional[List[Any]]:
        if n <= 0:
            return None
        res = []
        for _ in range(n):
            res.append(func())
        return res

    def get_backn(self, n: int) -> Optional[List[Any]]:
        return self._get_n(n, self.get_back)

    def get_frontn(self, n: int) -> Optional[List[Any]]:
        return self._get_n(n, self.get_front)

    def _add_back(self, elm: Any) -> None:
        new_end = (self._end + 1) % self._size

        if self._overflow_cb and self._length + 1 > self._size:
            self._overflow_cb(self._data[new_end], self)
        self._data[new_end] = elm

        if self._length < self._size:
            if self._length + 1 > self._size:
                self._length = self._size
            else:
                self._length += 1
        self._end = new_end
        self._start = (self._size + self._end - self._length + 1) % self._size

    def add_back(self, *items: Any) -> int:
        for item in items:
            self._add_back(item)
        return self._length

    def _add_front(self, elm: Any) -> None:
        if self._overflow_cb and self._length + 1 > self._size:
            self._overflow_cb(self._data[self._end], self)
        self._data[(self._size + self._start - 1) % self._size] = elm

        if self._size - self._length - 1 < 0:
            self._end += self._size - self._length - 1
            if self._end < 0:
                self._end = self._size + (self._end % self._size)

        if self._length < self._size:
            if self._length + 1 > self._size:
                self._length = self._size
            else:
                self._length += 1
        self._start -= 1
        if self._start < 0:
            self._start = self._size + (self._start % self._size)

    def add_front(self, *items: Any) -> int:
        for item in items:
            self._add_front(item)
        return self._length

    def unshift(self, *items: Any) -> int:
        for item in reversed(items):
            self._add_front(item)
        return self._length

    def get_all(self) -> List[Any]:
        raise NotImplementedError()

    def clear(self) -> None:
        self._length = 0
        self._start = 0
        self._data = [None] * self._size
        self._end = self._size - 1


class TopicBuffer(RingBuffer):
    def __init__(self, size: int, callback: Optional[Callable[[Any], bool]] = None) -> None:
        super().__init__(size)
        self.callback = callback


class MessagePool:
    def __init__(self, size: int = 100) -> None:
        self.ws_client = None
        self.topic_map: Dict[str, TopicBuffer] = {}
        self.buffer_size = size
        self.publish_method: Optional[Callable[[str], None]] = None
        self.debug = False
        self.auto_subscribe = False
        self.default_callback: Optional[Callable[[Any], bool]] = None

    def setPublishMethod(self, func: Callable[[str], None]) -> None:
        self.publish_method = func

    @staticmethod
    def makeMessageStr(topic_name: str, data: Any) -> str:
        msg = {"hash": "TUTCPS", "topic": topic_name, "data": data}
        return json.dumps(msg)

    def publish(self, topic_name: str, data: Any) -> None:
        if self.publish_method:
            msg_str = MessagePool.makeMessageStr(topic_name, data)
            if self.debug:
                print(f"publish: {msg_str}")
            self.publish_method(msg_str)

    def getMessage(self, topic_name: str, timeout: int = -1) -> Any:
        _ = timeout
        if topic_name in self.topic_map:
            buf = self.topic_map[topic_name]
            return buf.get_front()
        return None

    def readLastMessage(self, topic_name: str) -> Any:
        if topic_name in self.topic_map:
            buf = self.topic_map[topic_name]
            return buf.index(-1)
        return None

    def subscribe(self, topic_name: str, callback: Optional[Callable[[Any], bool]] = None) -> None:
        if topic_name in self.topic_map:
            buf = self.topic_map[topic_name]
            if callback:
                buf.callback = callback
        else:
            self.topic_map[topic_name] = TopicBuffer(self.buffer_size, callback)

    def unsubscribe(self, topic_name: str) -> None:
        if topic_name in self.topic_map:
            del self.topic_map[topic_name]

    @property
    def subscribingTopicList(self) -> List[str]:
        return list(self.topic_map.keys())

    @property
    def subscribingTopicIter(self) -> Iterator[str]:
        return iter(self.topic_map.keys())

    def _receiveData(self, instr: str) -> None:
        msg = json.loads(instr)
        if self.debug:
            print(f"revc: {msg}")
        if msg.get("hash") == "TUTCPS":
            if "topic" in msg and "data" in msg:
                topic = msg["topic"]
                data = msg["data"]
                if topic in self.topic_map:
                    buf = self.topic_map[topic]
                    if self.debug:
                        print(f"cb_func: {buf.callback}")
                    if buf.callback and buf.callback(data):
                        if self.debug:
                            print("do not stack message to buffer")
                        return
                    if self.debug:
                        print("stack message to buffer")
                    buf.push([datetime.now(), data])
                else:
                    if self.debug:
                        print(f"receive not subscribing topic : {topic}")
                    if self.auto_subscribe:
                        self.subscribe(topic, self.default_callback)
                        self._receiveData(instr)
#
#
#
import ssl
import threading
from websocket import create_connection
class WebMessage:
    def __init__(self, addr: str, size: int = 100, loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
        self.addr = addr
        self.size = size
        self.connection = False
        self.ws_client = None
        self.msg_pool = MessagePool(self.size)
        self._recv_thread: Optional[threading.Thread] = None
        self._stop_event = threading.Event()
        self.connect()

    def connect(self) -> None:
        try:
            ssl_context = ssl.create_default_context()
            ssl_context.check_hostname = False
            ssl_context.verify_mode = ssl.CERT_NONE
            self.ws_client = create_connection(self.addr, sslopt={"cert_reqs": ssl.CERT_NONE})
            self.msg_pool.setPublishMethod(lambda instr: self.ws_client.send(instr))
            self.connection = True
            print("ws open")
            self._stop_event.clear()
            self._recv_thread = threading.Thread(target=self._recv_loop, daemon=True)
            self._recv_thread.start()
        except Exception as e:
            print(f"ws connection error: {e}")
            self.connection = False

    def _recv_loop(self) -> None:
        try:
            while self.connection and not self._stop_event.is_set():
                msg = self.ws_client.recv()
                if msg:
                    self.msg_pool._receiveData(msg)
        except Exception:
            print("ws error")
        finally:
            print("ws closed")
            self.connection = False

    def close(self) -> None:
        self.connection = False
        self._stop_event.set()
        if self.ws_client:
            self.ws_client.close()
        if self._recv_thread:
            self._recv_thread.join(timeout=2)
#>#
#>#
#>#
#>import asyncio
#>import websockets
#>class WebMessage:
#>    def __init__(self, addr: str, size: int = 100, loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
#>        self.addr = addr
#>        self.size = size
#>        self.connection = False
#>        self.ws_client = None
#>        self.msg_pool = MessagePool(self.size)
#>        self._recv_task: Optional[asyncio.Task] = None
#>        self._loop = loop or asyncio.get_event_loop()
#>        self._loop.create_task(self.connect())
#>
#>    async def connect(self) -> None:
#>        self.ws_client = await websockets.connect(self.addr)
#>        self.msg_pool.setPublishMethod(lambda instr: self._loop.create_task(self.ws_client.send(instr)))
#>        self.connection = True
#>        print("ws open")
#>        self._recv_task = self._loop.create_task(self._recv_loop())
#>
#>    async def _recv_loop(self) -> None:
#>        try:
#>            async for msg in self.ws_client:
#>                self.msg_pool._receiveData(msg)
#>        except Exception:
#>            print("ws error")
#>        finally:
#>            print("ws closed")
#>            self.connection = False
#>
#>    async def close(self) -> None:
#>        if self.ws_client:
#>            await self.ws_client.close()
#>        if self._recv_task:
#>            await self._recv_task
#>

class BrowserMessage:
    def __init__(self, hostwindow: Any, size: int = 100) -> None:
        self.size = size
        self.hostwindow = hostwindow
        self.connection = False
        self.connect()

    def connect(self) -> None:
        raise NotImplementedError("BrowserMessage has no direct Python equivalent.")
